hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCancelAirdropHandler.java

```diff
@@ -0,0 +1,140 @@
+        }
+    }
+
+    /**
+     * Verifies that the sender account exists and ensures the account's key is valid and required for the transaction.
+     *
+     * @param senderId The AccountID of the specified sender whose key needs to be validated.
+     * @param context The PreHandleContext providing transaction context.
+     * @param accountStore The store to access readable account information.
+     * @throws PreCheckException If the sender's account is immutable or the sender's account ID is invalid.
+     */
+    private void verifyAndRequireKeyForSender(
+            @NonNull final AccountID senderId,
+            @NonNull final PreHandleContext context,
+            @NonNull final ReadableAccountStore accountStore)
+            throws PreCheckException {
+        final var senderAccount = accountStore.getAliasedAccountById(senderId);
+        validateTruePreCheck(senderAccount != null, INVALID_ACCOUNT_ID);
+
+        // If the sender account is immutable, then we throw an exception.
+        final var key = senderAccount.key();
+        context.requireKey(key);
+    }
+
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn, "Transaction body cannot be null");
+        final var op = txn.tokenCancelAirdropOrThrow();
+
+        validateFalsePreCheck(op.pendingAirdrops().isEmpty(), EMPTY_PENDING_AIRDROP_ID_LIST);
+        validateFalsePreCheck(
+                op.pendingAirdrops().size() > MAX_ALLOWED_PENDING_AIRDROPS_TO_CANCEL, MAX_PENDING_AIRDROP_ID_EXCEEDED);
+        final var uniquePendingAirdrops = new HashSet<PendingAirdropId>();
+        for (final var airdrop : op.pendingAirdrops()) {
+            if (!uniquePendingAirdrops.add(airdrop)) {
+                throw new PreCheckException(PENDING_AIRDROP_ID_REPEATED);
+            }
+            validateAccountID(airdrop.receiverId(), null);
+            validateAccountID(airdrop.senderId(), null);
+
+            if (airdrop.hasFungibleTokenType()) {
+                final var tokenID = airdrop.fungibleTokenType();
+                validateTruePreCheck(tokenID != null && !tokenID.equals(TokenID.DEFAULT), INVALID_TOKEN_ID);
+            }
+            if (airdrop.hasNonFungibleToken()) {
+                final var nftID = airdrop.nonFungibleTokenOrThrow();
+                validateTruePreCheck(nftID.tokenId() != null, INVALID_NFT_ID);
+                validateTruePreCheck(nftID.serialNumber() > 0, INVALID_TOKEN_NFT_SERIAL_NUMBER);
+            }
+        }
+    }
+
+    @Override
+    public void handle(@NonNull final HandleContext context) throws HandleException {
+        var tokensConfig = context.configuration().getConfigData(TokensConfig.class);
+        validateTrue(tokensConfig.cancelTokenAirdropEnabled(), NOT_SUPPORTED);
+    }
+
+    @NonNull
+    @Override
+    public Fees calculateFees(@NonNull final FeeContext feeContext) {
+        return Fees.FREE;
```

Comment:
nit: We should probably throw UnsupportedOperationException when something is yet to be implemented. It is possible we forget implementing this, if we leave as Fees.FREE. Need  not be changed here, just for future PRs 🙂

Reply:
This is deleted and the `feeCalculation` is implemented in the related PR: https://github.com/hashgraph/hedera-services/pull/14381